<h2 align='center' style='color:purple'>Outlier detection and removal using z-score and standard deviation in python pandas</h2>

PERCENTILE  
STD DEV -- ZSCORE
IQR

In [162]:
import pandas as pd
from matplotlib import pyplot as plt

In [163]:
df = pd.read_csv("heights.csv")
df.sample(5)

,gender,height
1987,Male,65.478267
4478,Male,65.566101
5800,Female,66.258258
6054,Female,65.476903
2383,Male,71.505206


In [168]:
df.height.mean()

66.3675597548656

In [169]:
df.height.std()

3.847528120795573

Here the mean is 66.37 and standard deviation is 3.84. 

<h3 style='color:blue'>(1) Outlier detection and removal using 3 standard deviation</h3>

One of the ways we can remove outliers is remove any data points that are beyond **3 standard deviation** from mean. Which means we can come up with following upper and lower bounds

In [170]:
upper_limit = df.height.mean() + 3*df.height.std()
upper_limit


77.91014411725232

In [171]:
lower_limit = df.height.mean() -3*df.height.std()
lower_limit

54.824975392478876

Here are the outliers that are beyond 3 std dev from mean

In [172]:
df[(df.height>upper_limit) | (df.height<lower_limit)]

,gender,height
994,Male,78.095867
1317,Male,78.462053
2014,Male,78.998742
3285,Male,78.528210
3757,Male,78.621374
6624,Female,54.616858
9285,Female,54.263133


Above the heights on higher end is **78 inch** which is around **6 ft 6 inch**. Now that is quite unusual height. There are people who have this height but it is very uncommon and it is ok if you remove those data points.
Similarly on lower end it is **54 inch** which is around **4 ft 6 inch**. While this is also a legitimate height you don't find many people having this height so it is safe to consider both of these cases as outliers

**Now remove these outliers and generate new dataframe**

In [173]:
df_no_outlier_std_dev = df[(df.height<upper_limit) & (df.height>lower_limit)]
df_no_outlier_std_dev.head()

,gender,height
0,Male,73.847017
1,Male,68.781904
2,Male,74.110105
3,Male,71.730978
4,Male,69.881796


Above shows original dataframe data 10000 data points. Out of that we removed 7 outliers (i.e. 10000-9993) 